docs: include full _copier_conf attributes list

wgordon17 opened the pull request.

wgordon17 commented:
Here's the funny thing, I hadn't even _noticed_ the introduction sentence with

```
_copier_conf includes a representation of the current Copier [Worker](https://copier.readthedocs.io/en/stable/reference/main/#copier._main.Worker) object
```

until I went back and looked _after_ I had found the `Attributes` table. Now that I _know_ it's there, this PR is really just intended to give some additional clarity (feel free to close as well if you feel that the extra `Note` shouldn't be necessary ¯\\\_(ツ)_/¯ )

sisp commented:
Thanks for aspiring to improve the docs! :bow: I agree that this the render context isn't documented as well as it could be. There's only one concern I have with this addition: `copier.main` will become an internal-only module and `copier.main.Worker` will become an internal-only class. So the API reference will become unavailable. IMO, we should model public interfaces (like the `copier.yml` content specification and the render context) with Pydantic to have declarative and explicit representations as single source of truth of information, static typing, runtime validation, and (exportable) JSON schema. I've resumed work on this long-standing idea of mine, but it's a quite massive change. In the meantime, how about documenting the render context explicitly and manually? We'll need to keep it synchronous with the actual implementation,, but changes are quite infrequent.

wgordon17 commented:
Hey @sisp, thanks for the review. If y'all are good with ensuring that the docs are just always in-sync, I'm _absolutely_ happy to help copy over the data. Let me do a quick fix, and I'll let you give it another review 😊

wgordon17 commented:
@sisp I think this should be ready for a review now. Direct link to docs preview: https://copier--2127.org.readthedocs.build/en/2127/creating/#_copier_conf

A few things that I adjusted (if you're OK with that)
- Alphabetized the attributes to ideally make it easier to skim, versus the previous order matching `_main.py`
- Included `Defaults` on the rows where I could easily see what the defaults were in `_main.py`
- Added a `NOTE` in `_main.py` to hopefully provide a reminder to folks that any changes there need to _also_ be reflected on the, now static, Attributes table in docs

wgordon17 commented:
@sisp Just checking if you're good with the changes?

sisp commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
 
-`_copier_conf` includes a representation of the current Copier
-[Worker][copier.main.Worker] object, also slightly modified:
+`_copier_conf` includes a slightly modified representation of the current Copier Worker
+object:
 
 -   It only contains JSON-serializable data.
 -   You can serialize it with `{{ _copier_conf|to_json }}`.
 -   ⚠️ It contains secret answers inside its `.data` key.
 -   Modifying it doesn't alter the current rendering configuration.
 
```
How about deleting this entire snippet, as referring to the `Worker` class and the relationship between the render context and the `Worker` class is an implementation detail?

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `pretend`          | `bool`                     | When `True`, produce no real rendering.<br>See the [`pretend`][] setting for related information.                                                                                                                                                          |
+| `quiet`            | `bool`                     | When `True`, disable all output.<br>See the [`quiet`][] setting for related information.                                                                                                                                                                   |
+| `sep`              | `str`                      | The operating system-specific directory separator.                                                                                                                                                                                                         |
+| `settings`         | `pydantic.BaseModel`       | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                 |
```
@sisp Thoughts on the content for this one?

sisp replied:
How about referring to the [`copier.settings.Settings`](https://github.com/copier-org/copier/blob/e2ce0d4f13f439c6d9467fc4bf0e0af5f6a60989/copier/settings.py#L21) model class via `[copier.settings.Settings][]` (thanks to mkdocstrings)?

```suggestion
| `settings`         | [copier.settings.Settings][] | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                                                                                                                                                         |
```

(The other rows will need re-formatting because of the wider column ...)

sisp replied:
Is the problem with `[copier.settings.Settings][]` that it doesn't get rendered as inline code? If so, I imagine ``[`copier.settings.Settings`][]`` or ``[`copier.settings.Settings`][copier.settings.Settings]`` should work – not tested though, as I'm commenting from mobile right now.

wgordon17 replied:
Yup, _perfect_! 👌 Which do you prefer `Settings` (A)  vs `copier.settings.Settings` (B) as the display value?

A ![SCR-20250607-qqea](https://github.com/user-attachments/assets/47b5019d-5781-4fbf-94c7-584ddada9f2e)

B ![SCR-20250607-qqrn](https://github.com/user-attachments/assets/e5bb238c-72c5-4e5b-bf87-deed7ad9f5c4)

sisp replied:
I think (B) is better because it's more explicit.

wgordon17 replied:
Great, since (B) is what's currently in the PR. I'll mark this one resolved then 👍

wgordon17 commented on docs/creating.md:
```diff
@@ -127,8 +127,8 @@ Attributes:
 | `user_defaults`    | `dict[str, Any]`           | Specified user defaults that may override a template's defaults during question prompts.                                                                                                                                                                   |
-| `vcs_ref`          | `str | None`               | The VCS tag/commit of the template, `None` if the template is not VCS-tracked.<br>See the [`vcs_ref`][] setting for related information.                                                                                                                   |
-| `vcs_ref_hash`     | `str | None`               | The VCS commit hash of the template, `None` if the template is not VCS-tracked.                                                                                                                                                                            |
+| `vcs_ref`          | `str                       | None`                                                                                                                                                                                                                                                      | The VCS tag/commit of the template, `None` if the template is not VCS-tracked.<br>See the [`vc ...
```
@sisp we might have to disable `autoformat` for this PR (not sure if you can). Leaving the `\` in `str \| None` caused the `\` to actually _print_ in the assembled docs, however removing it causes `autoformat` to think that it's a column separator and add extra whitespace which then breaks the table layout 🤦‍♂️

sisp replied:
Mhhh, not the first time that Prettier is causing some problems with Markdown formatting. How about using HTML and entity escaping in this particular case?

```suggestion
| `vcs_ref`          | <code>str &vert; None</code>                                                                                                                                                                                                                                                            | The VCS tag/commit of the template, `None` if the template is not VCS-tracked.<br>See the [`vcs_ref`][] setting for related information. |
```

Not great, but perhaps better than disabling Prettier altogether for the table.

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `pretend`          | `bool`                     | When `True`, produce no real rendering.<br>See the [`pretend`][] setting for related information.                                                                                                                                                          |
+| `quiet`            | `bool`                     | When `True`, disable all output.<br>See the [`quiet`][] setting for related information.                                                                                                                                                                   |
+| `sep`              | `str`                      | The operating system-specific directory separator.                                                                                                                                                                                                         |
+| `settings`         | `pydantic.BaseModel`       | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                 |
```
```suggestion
| `settings`         | `[Settings][copier.settings.Settings]`       | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                 |
```

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `pretend`          | `bool`                       | When `True`, produce no real rendering.<br>See the [`pretend`][] setting for related information.                                                                                                                                                          |
+| `quiet`            | `bool`                       | When `True`, disable all output.<br>See the [`quiet`][] setting for related information.                                                                                                                                                                   |
+| `sep`              | `str`                        | The operating system-specific directory separator.                                                                                                                                                                                                         |
+| `settings`         | `pydantic.BaseModel`         | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                 |
```
```suggestion
| `settings`         | `[Settings][copier.settings.Settings]`         | Specified settings that control copier's behavior.<br>See the [`settings`][] page for related information.                                                                                                                                                 |
```

wgordon17 commented:
@sisp 2 questions/thoughts, looking at 

![image](https://github.com/user-attachments/assets/aa38b3de-1765-4439-880a-ad23503acb02)

1. Why is `Linux` capitalized while the others are lowercase? Could/should that be normalized? 🤔 Likely a breaking change 😞 
2. Do you want to use `&nbsp;` to keep everything on one-line? Would definitely make the column wider 😞 Not sure if you have a preference ¯\\\_(ツ)_/¯

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `src_path`         | `Path`                                                        | The absolute path to the (cloned/downloaded) template on disk.                                                                                                                                                                                             |
+| `unsafe`           | `bool`                                                        | When `True`, allow usage of unsafe templates.<br>See the [`unsafe`][] setting for related information.                                                                                                                                                     |
+| `use_prereleases`  | `bool`                                                        | When `True`, `vcs_ref`/`vcs_ref_hash` may refer to a prerelease version of the template.<br>See the [`use_prereleases`][] setting for related information.                                                                                                 |
+| `user_defaults`    | `dict[str, Any]`                                              | Specified user defaults that may override a template's defaults during question prompts.                                                                                                                                                                   |
```
Should we point to https://copier--2127.org.readthedocs.build/en/2127/settings/#user-defaults? What would that be, `See the [User defaults][settings.user-defaults] setting for related information.`

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `exclude`          | `Sequence[str]`                                               | Specified additional [file exclusion patterns][patterns-syntax].<br>See the [`exclude`][] setting for related information.                                                                                                                                 |
+| `os`               | <code>Literal["Linux", "macos", "windows"] &vert; None</code> | The detected operating system, `None` if it could not be detected.                                                                                                                                                                                         |
+| `overwrite`        | `bool`                                                        | When `True`, overwrite files that already exist, without asking.<br>See the [`overwrite`][] setting for related information.                                                                                                                               |
+| `pretend`          | `bool`                                                        | When `True`, produce no real rendering.<br>See the [`pretend`][] setting for related information.                                                                                                                                                          |
```
```suggestion
| `pretend`          | `bool`                                                        | When `True`, perform no real rendering.<br>See the [`pretend`][] setting for related information.                                                                                                                                                          |
```
?

sisp replied:
Sure, but we should synchronize the descriptions and the `Worker` class attributes in a separate PR – wherever it makes sense.

wgordon17 commented on docs/creating.md:
```diff
@@ -95,27 +95,40 @@ suitable to [autoupdate your project safely][the-copier-answersyml-file]:
+| `context_lines`    | `PositiveInt`                                                 | Lines of context to consider when solving conflicts in updates.<br>See the [`context_lines`][] setting for related information.                                                                                                                            |
+| `data`             | `dict[str, Any]`                                              | Answers to the questionnaire, defined in the template, provided via CLI (`-d,--data`) or API (`data`).<br>See the [`data`][] setting for related information.<br>⚠️ May contain secret answers.                                                            |
+| `defaults`         | `bool`                                                        | When `True`, use default answers to questions.<br>See the [`defaults`][] setting for related information.                                                                                                                                                  |
+| `dst_path`         | `Path`                                                        | Destination path where to render the subproject.<br>⚠️ When [updating a project][updating-a-project], it may be a temporary directory, as Copier's update algorithm generates fresh copies using the old and new template versions in temporary locations. |
```
```suggestion
| `dst_path`         | `Path`                                                        | Destination path where to render the subproject.<br>⚠️ When [updating a project][updating-a-project], this may be a temporary directory, as Copier's update algorithm generates fresh copies using the old and new template versions in temporary locations. |
```

wgordon17 commented:
Ok, threw a few more "consistency" suggestions in here as well, WDYT?

sisp reviewed (commented):
> 1. Why is `Linux` capitalized while the others are lowercase? Could/should that be normalized? 🤔 Likely a breaking change 😞 

OMG, that's totally my error. I was typing on mobile and it seems typing correction capitalized it without me noticing. It should be lowercase of course. Thanks for catching this! :1st_place_medal:

> 2. Do you want to use `&nbsp;` to keep everything on one-line? Would definitely make the column wider 😞 Not sure if you have a preference ¯\\\_(ツ)_/¯ 

It does look better without the line wrap, but then the description column might get squeezed too much. I'd say, your call. We can always change it later. Thanks for paying attention to such details though! :bow:

wgordon17 commented:
@sisp Odd, it doesn't seem to be respecting my `&nbsp;`...oh well, for something we're not entirely sold on, it's probably not worth debugging ¯\\\_(ツ)_/¯ I've lowercased `linux`...perhaps it's finally time to ship it? 🚀

sisp reviewed (approved):
Looks great! :star: Thank you very much for working so diligently on this PR! :bow: Let's merge it. :wink:
